# Global Exploration missing values

In [12]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import outlier_removal as out
from sklearn import linear_model
df=pd.read_csv('TrainSample.csv',sep=';' )
from  numpy import nan

## Analysis by Lines

In [13]:
missing=df.isnull().as_matrix()
count_na_lines=[[i,np.sum(missing[i,:])] for i in range(len(missing))]
t=pd.DataFrame(count_na_lines)
t.columns = ['line','nb of nas per row']
t=t.groupby('nb of nas per row')['line'].count()
t=t.to_frame()
t.columns = ['amount of rows']
t


amount of rows
nb of nas per row                
0                             315
1                              41
2                               2
3                               4
5                               1
18                             73
19                             17
20                              5
21                              1
22                              1

Above we can see that we have two big groups of rows (records).We define:

Group1: less than 5    missing values. 363 members

Group2: more than 18 missing values. 97 members

### seperate the two groups

In [14]:
l1=[i for i in range(len(df)) if count_na_lines[i][1]<=6 ]
l2=[i for i in range(len(df)) if count_na_lines[i][1]>6 ]
group1=df.iloc[l1]
group2=df.iloc[l2]
group2

API    Surf_X    Surf_Y Date_Drilling Date_Completion Date_Production  \
5    382 -2.315616  1.968080    23/11/2013      13/03/2014      12/03/2014   
10   496 -2.275067  1.952019    11/10/2013      15/03/2014      10/03/2014   
11   534 -1.090920  2.077769    19/02/2014      09/10/2014      09/10/2014   
12   173 -1.002430  1.006312    27/10/2014      23/09/2015      25/09/2015   
20   450  2.869421 -1.004912    13/12/2013      23/07/2014      01/07/2014   
24   704 -1.011015  0.634942    25/03/2014      19/08/2014      19/08/2014   
31   362 -1.739415  2.327785           NaN      25/10/2014      25/10/2014   
44   461 -2.285832  1.934415    25/10/2013      14/03/2014      12/03/2014   
55   407 -2.398956  1.033041    19/12/2014      28/03/2015      15/05/2015   
61   495 -1.282817  0.709072    29/05/2013      16/10/2013      01/09/2013   
64   136 -1.788851  2.232945    06/01/2014      17/06/2014      17/06/2014   
67   767 -1.896971  1.958564    26/01/2014      09/10/2014      04/10/2014   
69   339 -2.034751  1.549856    24/04/2013      06/11/2013      28/03/2014   
70   649 -2.124975  2.317977    10/10/2014      14/01/2015      29/04/2015   
71   442 -1.211965  1.142066    02/09/2014             NaN      01/12/2015   
74   479 -1.328484  1.446556    20/08/2012      24/04/2013      01/04/2013   
81    97  0.192482 -1.734502    07/05/2012      14/08/2012      14/08/2012   
83   571 -0.367150  1.236809    22/12/2013      04/09/2014      28/08/2014   
86     4 -1.296250  1.119187    12/05/2013             NaN      01/09/2013   
90    53  0.481610 -1.171537    17/11/2013      20/08/2014      20/08/2014   
91   129 -2.301258  1.952937    07/11/2013      14/03/2014      12/03/2014   
92   491  2.832703 -1.063547           NaN      28/05/2015      12/04/2015   
95   711 -2.318824  1.773417           NaN      21/03/2015      21/03/2015   
98    36 -1.518191  0.725582    07/02/2014      19/06/2014      19/06/2014   
101  639 -1.770310  2.250490    27/11/2013      17/06/2014      21/06/2014   
105  564 -2.097495  3.136585    23/09/2014      19/04/2015      17/04/2015   
110  514 -2.307560  1.776992    06/07/2014      25/03/2015      25/03/2015   
115  270  2.845274 -1.048405    25/01/2015      25/05/2015      12/04/2015   
118  551  0.488774 -1.166965    05/03/2014      20/08/2014      20/08/2014   
121  660 -0.985747  0.647642    03/01/2015      27/07/2015      29/09/2015   
..   ...       ...       ...           ...             ...             ...   
276   87 -0.622279  0.662140    17/02/2014      25/08/2014      27/08/2014   
278  724 -0.611285  0.667669    05/02/2014      21/08/2014      18/08/2014   
288  781 -1.016673  0.998536    11/10/2014      20/09/2015      22/09/2015   
289  287 -1.221547  1.127744    11/08/2014             NaN      01/12/2015   
318  666 -2.048340  1.530376    10/05/2013      16/09/2013      19/02/2014   
325   72 -2.410459  0.999044    26/11/2014      28/03/2015      15/05/2015   
331  614 -1.257611  1.142633           NaN             NaN      01/01/2016   
332  618 -2.421556  1.463496    14/06/2014      20/11/2014      20/11/2014   
337   93 -2.161319  2.323233    21/08/2014      05/01/2015      20/04/2015   
339  706 -0.655250  0.645493    28/03/2014      29/08/2014      01/08/2014   
353  226 -0.878457  1.886369           NaN             NaN      10/06/2016   
358  484 -1.494127  0.870402    27/05/2013      12/12/2013      21/08/2013   
359  203 -0.625207  0.689260    11/04/2014      25/09/2014      01/09/2014   
373  166 -1.248341  0.963835    13/04/2014      27/11/2014      01/06/2015   
374  527 -1.266321  1.060317    25/07/2014             NaN      01/12/2015   
379  664 -2.083481  3.143658    25/08/2014      11/03/2015      20/04/2015   
384  218 -0.901856  1.874568    13/11/2015             NaN      06/06/2016   
385  608 -1.724736  2.754117           NaN      23/09/2014      23/09/2014   
388  732 -2.017228  1.551536    15/08/2013      21/12/2013      19/12/2013   
391  537 -0.426769  1.18278

### Analyze group 2
 We would like to know whether the missing values in group2 are in the same columns or each row has the missing values in different columns. A look into the data frame of group2 (above) confirms, that for group2, there is a set of 18 columns that is completely blank. We count the number of nans for each column of group2:

In [15]:
missing2=group2.isnull().as_matrix()
count_na_columns=[[i,np.sum(missing2[:,i])] for i in range(len(missing2[0]))] # substract data frame index
s=pd.DataFrame(count_na_columns)
s.columns = ['column','nb of nans in column']
s=s.groupby('nb of nans in column')['column'].count()
s=s.to_frame()
s.columns = ['amount of columns']
s['amount of columns'][0]=s['amount of columns'][0]-2  # data frame index given by python and API do not count 
s

amount of columns
nb of nans in column                   
0                                    23
1                                     1
7                                     1
13                                    2
97                                   18

We can confirm, that there is a set of 18 columns, which have all the 97 rows of group 2 missing.
For completeness, 
#### finish

## Analyis by columns
We count for each column of the whole dataset, how many missing values it has.

In [16]:
count_na_columns=[[i,np.sum(missing[:,i])] for i in range(len(missing2[0]))] # substract data frame index
s=pd.DataFrame(count_na_columns)
s.columns = ['column','nb of nans in column']
s=s.groupby('nb of nans in column')['column'].count()
s=s.to_frame()
s.columns = ['amount of columns']
s['amount of columns'][0]=s['amount of columns'][0]-2  # data frame index given by python and API do not count 
s

amount of columns
nb of nans in column                   
0                                    23
1                                     1
9                                     1
14                                    2
97                                    5
98                                    7
99                                    3
103                                   1
105                                   1
128                                   1

We have 23 complete features. We clearly have a jump in the above table between nb_na= 14 and nb_na = 97. having 14 missing values out of 460 can still be considered  a minor problem and estimation of those missing values will not introduce to much error in the model. On the other hand, having more than 97 out of 460 values missing is a bigger problem. We should not estimate 97 missing values in one column. 
We also devide the columns into two groups

column_group1:  less then 14 missing values with  29 members    ( including target variables )

column_group2. more than 97 missing values. with  18 


## seperate the two column groups
we split the data frame by the two column groups and name it data1 and data2 respectively.

In [17]:
k1=[df.columns[i] for i in range(len(df.columns)) if count_na_columns[i][1]<=14 ]
k2=[df.columns[i] for i in range(len(df.columns)) if count_na_columns[i][1]>14 ]
data1=df[k1]
data2=df[k2]


### Analyze Cgroup2
We want to confirm that in column-group2, the missing values are in the same lines. Hence we look at data 2 and count for each row the number of missing values.

In [18]:
missing3=data2.isnull().as_matrix()
count_na_lines=[[i,np.sum(missing3[i,:])] for i in range(len(missing3))]
t=pd.DataFrame(count_na_lines)
t.columns = ['line','nb of nas per row']
t=t.groupby('nb of nas per row')['line'].count()
t=t.to_frame()
t.columns = ['amount of rows']
t


amount of rows
nb of nas per row                
0                             317
1                              40
2                               2
3                               3
5                               1
18                             97

We can confirm, that there is 97 lines, that are empty in all the column group 2. These 97 lines correspond to the row-group two. Hence we get the same result as for the analyis by rows.

Finish the smaller groups.

## Detailed Analysis of columns
check whether the columns with more than 100 missing vals...
he columns with more than than 100 missing values are:

ShutInPressureInitilal  (103)
Frac Gradient           (105)
ShutInPressureInitilal  (128)

We already know that they have 97 missig lines in common (group). So we are going to focus on the other lines now.

As the two 'ShutInPressure' are highly correlated, we can reduce the number of missing vals of both of them.
In group1, they have only 1 missing value in common, so we will be able to reduce the number of missing values of each of them to 98.

In [40]:
missing_shut_in=np.logical_and(np.isnan(df['ShutInPressure_Fil (KPa)']),np.isnan (df['ShutInPressure_Initial (KPa)']))
group1[missing_shut_in]

API    Surf_X    Surf_Y Date_Drilling Date_Completion Date_Production  \
186  636  0.889209 -1.297541    28/03/2011      28/06/2011      24/07/2011   

     Lateral_Length (ft)  Depth_TVD_PPLS (ft)  Erosion_PPLS (ft)  \
186              0.01022             1.167945           0.100313   

     Pressure_PPLS (PSI)    ...      Min_Rate_Slurry (bpm)  \
186             0.686254    ...                   0.763242   

     ShutInPressure_Fil (KPa)  ShutInPressure_Initial (KPa)  ISIP (KPa)  \
186                       NaN                           NaN   -0.145505   

     Shot_Density (shots/ft)  Shot_Total  Proppant_per_ft (kg/ft)  \
186                 0.667339    3.151688                 1.409488   

     Stage_Spacing (ft)  GasCum360  OilCum360  
186            0.003925  -0.399839   0.451754  

[1 rows x 47 columns]

## Detailed Analysis of columns suite
check whether the columns with 1 to 14 missing vals...
1  missing val   'Date Production'
9  missing       'Pressure_PPLS (PSI)'
14 missing        Date_Drilling and Date_Completition

We want to whether the  Date_Drilling and Date_Completition have missing values in the same rows:


In [58]:
missing_date= np.logical_and(missing[:,3],missing[:,4])
df[missing_date]


API    Surf_X    Surf_Y Date_Drilling Date_Completion Date_Production  \
159  424 -1.665791 -1.109404           NaN             NaN             NaN   
238   24  0.131285  1.707435           NaN             NaN      14/04/2016   
331  614 -1.257611  1.142633           NaN             NaN      01/01/2016   
353  226 -0.878457  1.886369           NaN             NaN      10/06/2016   

     Lateral_Length (ft)  Depth_TVD_PPLS (ft)  Erosion_PPLS (ft)  \
159            -0.717850             3.497318          -0.183161   
238             0.709043            -2.676638          -0.331850   
331             0.656149            -0.705398           1.114258   
353             0.966668            -1.947109           1.313164   

     Pressure_PPLS (PSI)    ...      Min_Rate_Slurry (bpm)  \
159                  NaN    ...                        NaN   
238                  NaN    ...                   -0.44968   
331             0.295054    ...                        NaN   
353                  NaN    ...                        NaN   

     ShutInPressure_Fil (KPa)  ShutInPressure_Initial (KPa)  ISIP (KPa)  \
159                       NaN                           NaN         NaN   
238                  1.145678                      2.214658    1.187135   
331                       NaN                           NaN         NaN   
353                       NaN                           NaN         NaN   

     Shot_Density (shots/ft)  Shot_Total  Proppant_per_ft (kg/ft)  \
159                      NaN         NaN                      NaN   
238                 1.456334   -0.799406                -0.371036   
331                      NaN         NaN                      NaN   
353                      NaN         NaN                      NaN   

     Stage_Spacing (ft)  GasCum360  OilCum360  
159                 NaN  -1.241844  -0.694681  
238           -0.279994   0.622160  -0.846812  
331                 NaN  -0.655136   0.445087  
353                 NaN   0.531254  -0.698946  

[4 rows x 47 columns]

We see they have 4 missing values in common.